File: vhosts/blog/nuldoc-src/renderers/xml.ts
import { Element, forEachChild, Node, Text } from "../dom.ts";

export function renderXml(root: Node): string {
  return `<?xml version="1.0" encoding="utf-8"?>\n` + nodeToXmlText(root, {
    indentLevel: 0,
  });
}

type Context = {
  indentLevel: number;
};

type Dtd = { type: "block" | "inline" };

function getDtd(name: string): Dtd {
  switch (name) {
    case "feed":
    case "entry":
    case "author":
      return { type: "block" };
    default:
      return { type: "inline" };
  }
}

function isInlineNode(n: Node): boolean {
  if (n.kind === "text") {
    return true;
  }
  return getDtd(n.name).type === "inline";
}

function isBlockNode(n: Node): boolean {
  return !isInlineNode(n);
}

function nodeToXmlText(n: Node, ctx: Context): string {
  if (n.kind === "text") {
    if (n.raw) {
      return n.content;
    } else {
      return textNodeToXmlText(n);
    }
  } else {
    return elementNodeToXmlText(n, ctx);
  }
}

function textNodeToXmlText(t: Text): string {
  const s = encodeSpecialCharacters(t.content);

  // TODO: 日本語で改行するときはスペースを入れない
  return s.replaceAll(/\n */g, " ");
}

function encodeSpecialCharacters(s: string): string {
  return s.replaceAll(/&(?!\w+;)/g, "&amp;")
    .replaceAll(/</g, "&lt;")
    .replaceAll(/>/g, "&gt;")
    .replaceAll(/'/g, "&apos;")
    .replaceAll(/"/g, "&quot;");
}

function elementNodeToXmlText(e: Element, ctx: Context): string {
  let s = "";

  s += indent(ctx);
  s += `<${e.name}`;
  const attributes = getElementAttributes(e);
  if (attributes.length > 0) {
    s += " ";
    for (let i = 0; i < attributes.length; i++) {
      const [name, value] = attributes[i];
      s += `${name}="${encodeSpecialCharacters(value)}"`;
      if (i !== attributes.length - 1) {
        s += " ";
      }
    }
  }
  s += ">";
  if (isBlockNode(e)) {
    s += "\n";
  }
  ctx.indentLevel += 1;

  forEachChild(e, (c) => {
    s += nodeToXmlText(c, ctx);
  });

  ctx.indentLevel -= 1;
  if (isBlockNode(e)) {
    s += indent(ctx);
  }
  s += `</${e.name}>`;
  s += "\n";

  return s;
}

function indent(ctx: Context): string {
  return "  ".repeat(ctx.indentLevel);
}

function getElementAttributes(e: Element): [string, string][] {
  return [...e.attributes.entries()]
    .filter((a) => !a[0].startsWith("--"))
    .sort(
      (a, b) => {
        // Special rules:
        if (e.name === "link") {
          if (a[0] === "href" && b[0] === "rel") {
            return 1;
          }
          if (a[0] === "rel" && b[0] === "href") {
            return -1;
          }
          if (a[0] === "href" && b[0] === "type") {
            return 1;
          }
          if (a[0] === "type" && b[0] === "href") {
            return -1;
          }
        }
        // General rules:
        if (a[0] > b[0]) return 1;
        else if (a[0] < b[0]) return -1;
        else return 0;
      },
    );
}

File: services/nuldoc/lib/nuldoc/renderers/xml.rb
module Nuldoc
  class XMLRenderer
    BLOCK_ELEMENTS = %w[feed entry author].freeze

    def render(root)
      "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n#{node_to_xml(root, indent_level: 0)}"
    end

    private

    def inline_node?(node)
      return true if %i[text raw].include?(node.kind)

      !BLOCK_ELEMENTS.include?(node.name)
    end

    def block_node?(node)
      !inline_node?(node)
    end

    def node_to_xml(node, indent_level:)
      case node.kind
      when :text
        text_node_to_xml(node)
      when :raw
        node.html
      when :element
        element_node_to_xml(node, indent_level: indent_level)
      end
    end

    def text_node_to_xml(node)
      encode_special_characters(node.content).gsub(/\n */, ' ')
    end

    def encode_special_characters(s)
      s.gsub(/&(?!(?:\w+|#\d+|#x[\da-fA-F]+);)/, '&amp;')
       .gsub('<', '&lt;')
       .gsub('>', '&gt;')
       .gsub("'", '&apos;')
       .gsub('"', '&quot;')
    end

    def element_node_to_xml(element, indent_level:)
      s = +''

      s << indent(indent_level)
      s << "<#{element.name}"

      attributes = get_element_attributes(element)
      unless attributes.empty?
        s << ' '
        attributes.each_with_index do |(name, value), i|
          s << "#{name}=\"#{encode_special_characters(value)}\""
          s << ' ' if i != attributes.length - 1
        end
      end
      s << '>'
      s << "\n" if block_node?(element)

      child_indent = indent_level + 1
      element.children.each do |child|
        s << node_to_xml(child, indent_level: child_indent)
      end

      s << indent(indent_level) if block_node?(element)
      s << "</#{element.name}>"
      s << "\n"

      s
    end

    def indent(level)
      '  ' * level
    end

    def get_element_attributes(element)
      element.attributes
             .reject { |k, _| k.start_with?('__') }
             .sort { |a, b| compare_attributes(element.name, a, b) }
    end

    def compare_attributes(element_name, a, b)
      ak, = a
      bk, = b

      if element_name == 'link'
        return 1 if ak == 'href' && bk == 'rel'
        return -1 if ak == 'rel' && bk == 'href'
        return 1 if ak == 'href' && bk == 'type'
        return -1 if ak == 'type' && bk == 'href'
      end

      ak <=> bk
    end
  end
end
